changefeedccl: Add simplistic pushback mechanism.

miretskiy opened the pull request:
Add a simplistic pushback mechanism to changefeeds.  This mechanism
uses channel buffer, instead of the default memory buffer.
The effect is that if we are slow to read off of the channel
(perhaps due to slow sink, or very high request rate), we will
no longer be reading from the channel, which results in a pushback effect.

This is a simplistic implementation most likely has
higher CPU and memory allocation/deallocation costs.
The pushback mechanism will be improved upon in the future.
The default is to use memory buffer, but this mechanism can
be enabled via `changefeed.enable_pushback` setting.

Informs #63186
Informs #60697

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64584)
<!-- Reviewable:end -->

miretskiy commented:
I disabled TestChangefeedMonitoring test -- see #64583

Also, I don't have any dedicated tests for pushback -- but, see `configureMetamorphic` method -- I think we get all of the test coverage of the existing test.
If you think I should add dedicated pushback test (and, have ideas on how I could write one), let me know.

miretskiy commented:
Also note, I've tried to keep the scope of this change to absolute minimum -- we might want to backport it.

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64584)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64584#-MZq__ocARE1g787iQAD:-MZq__ocARE1g787iQAE:b-1st644) (no related file):*
It might be good to do a manual check of what this does to the performance of a moderately sized backfill.

Do you think we might want to be able to configure either the buffer size of this channel buffer or the existing eventC buffer (see physical_kv_feed.go)? My mild worry is that when you turn on this setting, I believe that means that there is currently at most 128 unhandled rangefeed messages before we fail the processor.  This seems potentially a bit low given the cluster sizes we've been discussing.

---


<!-- Sent from Reviewable.io -->

miretskiy commented:
Ouch.  This might be bad news.

On Tue, May 4, 2021 at 5:23 AM Steven Danna ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> *Reviewable <https://reviewable.io/reviews/cockroachdb/cockroach/64584>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on
> @ajwerner <https://github.com/ajwerner> and @miretskiy
> <https://github.com/miretskiy>)
>
> *a discussion
> <https://reviewable.io/reviews/cockroachdb/cockroach/64584#-MZq__ocARE1g787iQAD:-MZq__ocARE1g787iQAE:b-1st644>
> (no related file):*
> It might be good to do a manual check of what this does to the performance
> of a moderately sized backfill.
>
> Do you think we might want to be able to configure either the buffer size
> of this channel buffer or the existing eventC buffer (see
> physical_kv_feed.go)? My mild worry is that when you turn on this setting,
> I believe that means that there is currently at most 128 unhandled
> rangefeed messages before we fail the processor. This seems potentially a
> bit low given the cluster sizes we've been discussing.
> ------------------------------
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/64584#pullrequestreview-651041561>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVCQXYSZHGWQ6VQCLRDTL64KHANCNFSM44BIIRNQ>
> .
>

Commit:
changefeedccl: Issue a flush request when waiting for more memory.

The blocking buffer enables changefeeds to limit the amount of
ingested messages based on the available memory.  The memory required
to process an event is allocated once the event has been admitted,
and is released once the message has been emitted to the sink.

Some sinks (e.g. cloud storage sinks) buffer user configurable
number of events.  While the sink has buffered events, it does not
release memory.  Thus, it is possible that all of the previously ingested
events are buffered (but the amount of buffered data is below user
defined threshold), while we cannot ingest any more events because
of memory pushback.  This results in changefeeds being stuck.

This change allows the blocking buffer to signal the event consumer
that it should flush its data.  The signaling is accomplished via a new
type of event -- the flush event.  The event consumer, upon seeing such event
flushes its sink, thus releasing memory and making it possible to make
forward progress.

Fixes #70248

Release Justification: Fix to a newly added functionality to prevent stuck
changefeeds.

Release Notes: None.
